### PR TITLE
add: --last-known-good-release option

### PR DESCRIPTION
**Pull request by al-k21**
I guess this would be the first step.

Refs: https://github.com/nodejs/citgm/issues/407

##### Checklist

- [x] `npm test` passes
- [ ] documentation is changed or added
- [x] contribution guidelines followed [here](https://github.com/nodejs/citgm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by MylesBorins** [comment]
So the code looks like it should do what it says it does, but it would be nice to have some tests to ensure that is the case. This will also need documentation before it can land.

What do we do in the case where someone asks for LKGR but it is undefined? Should we be running any of the tests that are not a known good version?

Before you go ahead and do that I'd like for us to discuss what LKGR means. Since we test on so many different platforms + versions, how can we ensure that LKGR is meaningful. I'm also curious on how we expect to do ongoing updates with this. A tool to manually update the lookup would be useful, but again has the issue of platform + version inconsistencies.

To be clear, I think this can be useful, I'm just not 100% convinced that we should solve this in citgm as opposed to coming up with a solution in CI

**Comment by al-k21** [comment]
> So the code looks like it should do what it says it does, but it would be nice to have some tests to ensure that is the case. This will also need documentation before it can land.

Sure, I will add tests and documentation (not done it yet, as I wasn't sure this is going to land).

> What do we do in the case where someone asks for LKGR but it is undefined?

I guess we should use the latest from npm and warn the user that it is the case. ATM the code will use the latest version from npm.

> A tool to manually update the lookup would be useful, but again has the issue of platform + version inconsistencies.

A separate tool to update lookup sounds reasonable. I understand that the lookup will explode in size but I don't think this would be an issue.

> I'm just not 100% convinced that we should solve this in citgm as opposed to coming up with a solution in CI

Not sure what a `CI solution` could be...

**Comment by MylesBorins** [comment]
I would hold off on tests / doc until we have consensus on landing this.

A CI solution could be running the entire suite on the last version of node, and then again on the new version of node we are trying to test. then doing a DIFF against the results and only publishing the failures that were new for the new version

**Comment by gibfahn** [comment]
>A CI solution could be running the entire suite on the last version of node, and then again on the new version of node we are trying to test. then doing a DIFF against the results and only publishing the failures that were new for the new version

So something like https://github.com/nodejs/citgm/issues/232 😜 .

**Comment by gibfahn** [comment]
>A separate tool to update lookup sounds reasonable. I understand that the lookup will explode in size but I don't think this would be an issue.

I think it's not so much the size increase as the complexity increase that would be a problem.

File: bin/citgm-all.js [hunk 1]
```diff
@@ -38,6 +38,11 @@ const yargs = commonArgs(require('yargs'))
   .option('excludeTags', {
     type: 'array',
     description: 'Define which tags from the lookup to skip'
+  })
+  .option('last-know-good-release', {
+    alias: 'lkgr',
+    type: 'boolean',
+    description: 'Uses last know good release'
```
Comment (BethGriggs) [inline bin/citgm-all.js]: known

**Comment by mhdawson** [comment]
One advantage of this approach over the CI one is that it would be easier for outside people to run.

In terms of LKGR, today I assume we only run CiTGM on a single version and it either passes or fails.  Could we have each run of CiTGM write out a file that captures the versions run on and the pass fail status.  In the CI we'd them have to push that file for each platform to a common place and then those could be used in subsequent runs.  As opposed to LKGR you might request to use a specific baseline (one of the eariler runs) and it would only run on a particular platform it if it had passed before and thus would report any new failures.  It would have the drawback that one you have one failure on a platform you might miss subsequent ones.

**Comment by MylesBorins** [comment]
@mhdawson I like where this thought process is going

Building on it here is an idea for an implementation

* Dot File saved on system
* Maintain meta data for LKGR of a specific dependency per semver-major version in this dot-file
  - as this is a local system file it doesn't need to know anything about system
  - as this is a local system file we don't need to maintain this data in the lookup
* CITGM could save to this file by default, but only utilize it for reporting when given a specific flag
  - utilizing would mean installing that specific version rather than whatever the latest version is
* If a project has failed in a previous run and fails once again we can mark it as flaky rather than fail
* We can iterate this approach to allow named / explicit run comparison (this could be great for ABI testing)
* We can make a flag to pass in an arbitrary lookup of LKGRs, allowing us to inject specific versions to test

With this approach we get 0 ongoing maintenance while maintaining the benefits of keeping track of LKGR. As it is opt-in to utilize the data we will not be ignoring latest packages, and updating the lookup would be as simple as running citgm (something we can automate with nightlies).

**Comment by mhdawson** [comment]
I agree we can start with just saving on local system, although if we can to use in CI we'd have to figure out a mechanism to share the files across machines as there is no guarantee you would run on the same machine each time.  However, I think can figure that out in the CI after we have it working.

**Comment by MylesBorins** [comment]
TBH, we may not want to share between machines.

Each machine has it's own weirdness

On Mar 1, 2018 1:29 PM, "Michael Dawson" <notifications@github.com> wrote:

I agree we can start with just saving on local system, although if we can
to use in CI we'd have to figure out a mechanism to share the files across
machines as there is no guarantee you would run on the same machine each
time. However, I think can figure that out in the CI after we have it
working.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/nodejs/citgm/pull/551#issuecomment-369685390>, or mute
the thread

**Comment by MylesBorins** [comment]
This has not had any progress is over a year. Closing but please feel free to reopen

**Pull request closed without merging**